vms/avm/txs/mempool/mempool_test.go

```diff
@@ -13,196 +14,266 @@ import (
-		require.FailNow("should not have sent message to engine")
-	default:
-	}
+	tx0 := newTx(0, 32)
+	tx1 := newTx(1, 32)
+
+	require.NoError(mempool.Add(tx0))
+	require.NoError(mempool.Add(tx1))
 
-	for _, tx := range testTxs {
-		txID := tx.ID()
-		// tx not already there
-		require.False(mempool.Has(txID))
+	tx, exists := mempool.Peek()
+	require.True(exists)
+	require.Equal(tx, tx0)
 
-		// we can insert
-		require.NoError(mempool.Add(tx))
+	mempool.Remove([]*txs.Tx{tx0})
 
-		// we can get it
-		require.True(mempool.Has(txID))
+	tx, exists = mempool.Peek()
+	require.True(exists)
+	require.Equal(tx, tx1)
 
-		retrieved := mempool.Get(txID)
-		require.NotNil(retrieved)
-		require.Equal(tx, retrieved)
+	mempool.Remove([]*txs.Tx{tx0})
 
-		// tx exists in mempool
-		require.True(mempool.Has(txID))
+	tx, exists = mempool.Peek()
+	require.True(exists)
+	require.Equal(tx, tx1)
 
-		// once removed it cannot be there
-		mempool.Remove([]*txs.Tx{tx})
+	mempool.Remove([]*txs.Tx{tx1})
 
-		require.False(mempool.Has(txID))
-		require.Nil(mempool.Get(txID))
+	_, exists = mempool.Peek()
+	require.False(exists)
+}
 
-		// we can reinsert it again to grow the mempool
-		require.NoError(mempool.Add(tx))
+func TestIterate(t *testing.T) {
+	require := require.New(t)
+
+	mempool, err := New(
+		"mempool",
+		prometheus.NewRegistry(),
+		nil,
+	)
+	require.NoError(err)
+
+	var (
+		iteratedTxs []*txs.Tx
+		maxLen      = 2
+	)
+	addTxs := func(tx *txs.Tx) bool {
+		iteratedTxs = append(iteratedTxs, tx)
+		return len(iteratedTxs) < maxLen
 	}
+	mempool.Iterate(addTxs)
+	require.Empty(iteratedTxs)
+
+	tx0 := newTx(0, 32)
+	require.NoError(mempool.Add(tx0))
+
+	mempool.Iterate(addTxs)
+	require.Equal([]*txs.Tx{tx0}, iteratedTxs)
+
+	tx1 := newTx(1, 32)
+	require.NoError(mempool.Add(tx1))
+
+	iteratedTxs = nil
+	mempool.Iterate(addTxs)
+	require.Equal([]*txs.Tx{tx0, tx1}, iteratedTxs)
+
+	tx2 := newTx(2, 32)
+	require.NoError(mempool.Add(tx2))
+
+	iteratedTxs = nil
+	mempool.Iterate(addTxs)
+	require.Equal([]*txs.Tx{tx0, tx1}, iteratedTxs)
+
+	mempool.Remove([]*txs.Tx{tx0, tx2})
+
+	iteratedTxs = nil
+	mempool.Iterate(addTxs)
+	require.Equal([]*txs.Tx{tx1}, iteratedTxs)
+}
+
+func TestRequestBuildBlock(t *testing.T) {
+	require := require.New(t)
+
+	toEngine := make(chan common.Message, 1)
+	mempool, err := New(
+		"mempool",
+		prometheus.NewRegistry(),
+		toEngine,
+	)
+	require.NoError(err)
 
 	mempool.RequestBuildBlock()
 	select {
 	case <-toEngine:
+		require.FailNow("should not have sent message to engine")
 	default:
-		require.FailNow("should have sent message to engine")
 	}
 
-	mempool.Remove(testTxs)
+	tx := newTx(0, 32)
+	require.NoError(mempool.Add(tx))
 
 	mempool.RequestBuildBlock()
+	mempool.RequestBuildBlock() // Must not deadlock
 	select {
 	case <-toEngine:
-		require.FailNow("should not have sent message to engine")
 	default:
+		require.FailNow("should have sent message to engine")
 	}
-}
-
-func createTestTxs(count int) []*txs.Tx {
```

Comment:
This function _must_ have been a copy paste from somewhere... because it is massively overkill for the tests in this file